### PR TITLE
Fixed #30849 -- Added autodetector signals.

### PR DESCRIPTION
**Pull request by Skeen**
Added two new signals, sent during makemigrations, which allow
third-party applications to modify the change detection process, and
allows them to insert custom operations into automatically generated
migration files, based upon changes to the model state.

### CODE REVIEW

**Comment by Skeen** [comment]
I'm not sure how I can get rid of the documentation warnings for missing references:
```
docs/ref/migration-operations.txt:544: WARNING: py:data reference target not found: django.db.models.signals.post_autodetect
docs/releases/3.0.txt:312: WARNING: py:data reference target not found: django.db.models.signals.post_autodetect
```

**Comment by felixxm** [comment]
Closing per ticket.

**Comment by charettes** [comment]
Closing per ticket; a clear use case and wide spread community interest has not been demonstrated on the mailing list thread.

The auto-detector logic is not formalized and allowing injection of operations by third party than cannot be optimized through will likely cause the generation of inefficient migrations. Most of the `generate_*` methods generate inefficient operation sequences knowing that they'll be optimized any non-optimizable operation will break process.

To summarize, I'm not saying we should not eventually add this feature if there's wide spread interest in it but the limited provided motives doesn't outweigh the maintenance burden this would incur IMO.
